Add coupon link instead of dropdown in Cart and Checkout blocks

tarunvijwani opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
With this PR, we're changing the look and feel of the coupon code in both the Cart and Checkout blocks. We're removing the collapsed section and replacing it with a link.
<!-- Reference any related issues or PRs here -->

Fixes woocommerce/woocommerce-blocks#7890

## Changes in the PR
- Fixed the coupon code getting cleared when there is a validation error. 
- Replaced the collapsed section with a link.
- Fix unit tests
- Update the storybook.

Updating the e2e test cases in woocommerce/woocommerce-blocks#8022 


<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->

#### Accessibility

<!-- If you've changed or added any interactions, check off the appropriate items below. You can delete any that don't apply. Use this space to elaborate on anything if needed. -->

- [ ] I've tested using only a keyboard (no mouse)
- [ ] I've tested using a screen reader
- [ ] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [ ] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

### Screenshots

<!-- If your change has a visual component, add a screenshot here. A "before" screenshot would also be helpful. -->

| Before | After |
| ------ | ----- |
|![image](https://user-images.githubusercontent.com/11503784/209286969-60039ee0-0cf4-4adf-bb63-1235bc71ab7a.png)|![image](https://user-images.githubusercontent.com/11503784/209287005-42a8f487-38e3-4572-a60b-fd3da4d53656.png)|

### Testing

#### Automated Tests
* [x] Changes in this PR are covered by Automated Tests.
  * [x] Unit tests
  * [ ] E2E tests

#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

1. Checkout this branch
2. Go to the Edit Cart page, remove the Cart block and add it again. 
3. Go to the Edit Checkout page, remove the Checkout block and add it again. 
4. Visit your site front-end and add items to cart
5. Go to the Cart page, and confirm the `Apply a coupon code` link is visible.
6. Enter the coupon code, and confirm the coupon is getting applied. 
7. Go to the Checkout page, and confirm the `Apply a coupon code` link is visible.
6. Enter the coupon code, and confirm the coupon is getting applied. 
7. Test around this issue. 


##### Storybook

1. Checkout this branch
2. Run `npm run storybook:build`
3. Go to Storybook and confirm the new design is updated in Storybook. 

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental



### Changelog

> Enhancement: Replace the collapsed section for the coupon code with a link.

Commit:
Add coupon link instead of dropdown in Cart and Checkout blocks

Commit:
Merge branch 'trunk' into add/7890-coupon-code-link

Commit:
Merge branch 'trunk' into add/7890-coupon-code-link

Commit:
Fix state names used for displaying the link and coupon form

Commit:
Fix storybook for coupon block

Commit:
Fix unit test for coupon block

tarhi-saad commented on assets/js/base/components/cart-checkout/totals/coupon/index.tsx:
```diff
@@ -42,10 +36,11 @@ export interface TotalsCouponProps {
 }: TotalsCouponProps ): JSX.Element => {
 	const [ couponValue, setCouponValue ] = useState( '' );
+	const [ displayCouponForm, setdisplayCouponForm ] = useState( false );
+	const [ displayAddCouponLink, setdisplayAddCouponLink ] = useState( true );
```
@tarunvijwani, let's use `camelCase` for these state setters: `setDisplayCouponForm` & `setDisplayAddCouponLink`

Also, we don't need both of these states! We can use only one of them and get the same result. For example, let's choose this one:

`const [ displayCouponForm, setdisplayCouponForm ] = useState( false );`

When rendering our React component, we can use a ternary condition instead:

```JS
<>
  { displayCouponForm ? (
    // Our coupon form JSX elements...
  ) : (
    // Our coupon link JSX elements...
  ) }
</>
```

tarunvijwani replied:
Ah, it was the typo! Updated the code. Thank you!

tarhi-saad commented on assets/js/base/components/cart-checkout/totals/coupon/style.scss:
```diff
@@ -39,3 +39,9 @@
 	position: relative;
 }
+a.wc-block-components-totals-coupon-link {
+	text-decoration: none !important;
```
@tarunvijwani, I tested using the default theme (i.e., 2023), and the `!important` keyword wasn't necessary in my case! Could you check again on your end?

tarunvijwani replied:
If I remove important, it won't work for the storefront theme. Is there a better way to tackle this?

ralucaStan replied:
What was the reason this class was added, cause it might not be needed? That anchor should inherit the theme styles, just like [here](https://github.com/woocommerce/woocommerce-blocks/issues/8027#issuecomment-1376753360), and should work across the themes. For example, in TT2 the anchors are underlined so the line above would prevent the underline.

tarunvijwani replied:
@ralucaStan Yes, that makes sense. I'll remove the CSS. 

We would need this class name in our e2e test cases to select the element and click on it to display the coupon form. So I think we should keep the class as it is?

ralucaStan replied:
👍  sounds good.

tarhi-saad commented on assets/js/base/components/cart-checkout/totals/coupon/style.scss:
```diff
@@ -39,3 +39,9 @@
+a.wc-block-components-totals-coupon-link {
+	text-decoration: none !important;
+	cursor: pointer;
+	padding-left: 16px;
```
With this `padding-left` value, the text isn't aligned with the rest of other section titles:

<img width="334" alt="image" src="https://user-images.githubusercontent.com/14235870/209662214-ec204ef2-ec8c-4085-b445-df716f00a05f.png">

tarunvijwani replied:
Fixed it! 🙂

tarhi-saad commented on assets/js/blocks/cart/inner-blocks/cart-order-summary-block/edit.tsx:
```diff
@@ -33,10 +33,10 @@ export const Edit = ( { clientId }: { clientId: string } ): JSX.Element => {
 		[ 'woocommerce/cart-order-summary-subtotal-block', {}, [] ],
 		[ 'woocommerce/cart-order-summary-fee-block', {}, [] ],
 		[ 'woocommerce/cart-order-summary-discount-block', {}, [] ],
-		[ 'woocommerce/cart-order-summary-coupon-form-block', {}, [] ],
```
@tarunvijwani, did we change this to update the order of the coupon form block? In my test site, it's still positioned below the `Subtotal`/`Discount` sections!

tarunvijwani replied:
Are you on the Cart or the Checkout page? I should have changed the position on the Checkout page as well. Updated the code. Please re-add the block in the editor and test.

tarhi-saad commented on assets/js/data/cart/actions.ts:
```diff
@@ -237,8 +238,6 @@ export const applyCoupon =
 			// Re-throw the error.
 			throw error;
-		} finally {
-			dispatch.receiveApplyingCoupon( '' );
```
@tarunvijwani, what was the issue with dispatching the `receiveApplyingCoupon` action at this stage?

tarhi-saad replied:
I believe this change was needed to fix the following issue:

- Fixed the coupon code getting cleared when there is a validation error.

tarunvijwani replied:
Yes, the coupon code was getting cleared when there was a validation error.

tarhi-saad commented on assets/js/base/components/cart-checkout/totals/coupon/stories/index.tsx:
```diff
@@ -53,7 +50,6 @@ export const ErrorState: Story< TotalsCouponProps > = ( args ) => {
 	const { setValidationErrors } = useDispatch( VALIDATION_STORE_KEY );
 
 	setValidationErrors( { coupon: INVALID_COUPON_ERROR } );
-
```
@tarunvijwani, in `Storybook`, the `Error State` of the `Coupon` doesn't display an error:

| Before | After |
| ------ | ----- |
|<img width="969" alt="image" src="https://user-images.githubusercontent.com/14235870/209669121-fa777d12-1168-4e4c-a983-80acbc048f88.png">|<img width="988" alt="image" src="https://user-images.githubusercontent.com/14235870/209669245-f1faa116-b761-4151-b9f8-dbd52377cfaf.png">|

The reason for this issue is when we click on the `Add a coupon` link to show the coupon form, the `clearValidationErrors()` action is triggered! Which removes the error added by the `setValidationErrors` from this line! A simple solution would be to show the coupon form directly without clicking on the link. You can do that by putting back the argument `initialOpen: true`.

tarunvijwani replied:
I have removed the changes related to Storybook because of bug #8118. I'll open a separate PR for Storybook once the issue is resolved.

tarhi-saad reviewed (changes requested):
Hi, @tarunvijwani! 👋 Thank you for working on this PR! I've left some comments below! Please let me know if you have any questions! 🙌

ralucaStan commented on assets/js/base/components/cart-checkout/totals/coupon/index.tsx:
```diff
@@ -63,81 +58,89 @@ export const TotalsCoupon = ( {
+				// eslint-disable-next-line jsx-a11y/anchor-is-valid
+				<a
+					role="button"
+					href="#"
```
Nice work adapting this functionality. I noticed that the initial implementation was not accessible so I'm proposing to improve this in your implementation. If you would try to test this branch with the screen reader you will see that it is not a good experience.

We'll need to tweak this implementation to work well for screen readers as well.
Looking at the current implementation there are two issues:
- the anchor is not pointing to the hidden element; `href="#"` should be avoided, as it's not accessible. 
- the hidden element is not included in the page, making the whole area not screen reader friendly

Actually, the href of this link is the form (or the div container of the form, depending on what is hidden), so ideally the href should be "#id-of-the-from". Where <div id="id-of-the-from" ../>
The hidden content should be rendered, but not present on the page. I think there were some CSS classes in the project to hide content outside of the screen readers. When the user clicks the link the CSS class should be removed and the content should appear. 

We'd need to also make sure there is no issue with validation in this case, as in the form is visually hidden but still we run validation on it.

tarunvijwani replied:
That's actually a better approach. Thank you, I have made the changes! 👍🏻

Commit:
Fix css and accessibility

Commit:
Fix coupon block position on Checkout

tarunvijwani commented:
Thank you so much @tarhi-saad & @ralucaStan for the review comments. I have made the suggested changes. 🙂

Commit:
Update e2e test cases for coupon block (#8022)

* Update e2e test cases for coupon block

* Fix perfomance test for coupon block

* Fix double await typo

* Fix perfomance test for coupon block

tarunvijwani commented:
Merged the e2e test case changes(#8022) into the branch.

Commit:
Remove CSS

tarhi-saad reviewed (approved):
Thank you for the changes, @tarunvijwani! I tested it again! It looks good. Let's 🚢 it! 🙌

ralucaStan commented on assets/js/base/components/cart-checkout/totals/coupon/index.tsx:
```diff
@@ -63,30 +68,33 @@ export const TotalsCoupon = ( {
+			{ couponFormClass === 'screen-reader-text' && (
+				<a
+					role="button"
+					href="#wc-block-components-totals-coupon__form"
```
This href needs an element that has the id `wc-block-components-totals-coupon__form`, and I noticed that there is no element with this id, only the form has it as a class.

tarunvijwani replied:
I have added the ID to the element. 👍

ralucaStan commented on assets/js/base/components/cart-checkout/totals/coupon/index.tsx:
```diff
@@ -63,30 +68,33 @@ export const TotalsCoupon = ( {
-				/>
-			}
-		>
+					onClick={ (
```
This solution works but there is room for improvement and the opportunity to make it cleaner.

1. It's preferable to pass one function to the onClick prop, and handle the logic in that function.
```
onClick ={handleCouponAnchorClick}
```

```
handleCouponAnchorClick = (e:React.MouseEvent< HTMLAnchorElement, MouseEvent >) => {
e.preventDefault();
...
}

tarunvijwani replied:
Thank you for the suggestion. I have updated the code. 👍

ralucaStan commented on assets/js/base/components/cart-checkout/totals/coupon/index.tsx:
```diff
@@ -95,7 +103,12 @@ export const TotalsCoupon = ( {
 			>
-				<div className="wc-block-components-totals-coupon__content">
+				<div
+					className={
```
When dealing with changing classes we use the `classnames` package. 
In order to use it you'd have to change the logic you are using. Using string values to detect a state of an element is error-prone (classes usually change, string comparison is tricky), so it's better to work with boolean values. 
I recommend walking away from the `couponFormClass` and introducing `isCouponFormHidden` as a way to check the state of the component. The latter has the advantage of being more explicit about the functionality it handles, it's clear that the form is sometimes hidden and it depends on the value of `isCouponFormHidden`.

Class change should be handled as follows: 

```
import classnames from 'classnames';
...
const formWrapperClass = classnames(
		'wc-block-components-totals-coupon__content',
		{
			'screen-reader-text': isCouponFormHidden,
		}
	);
... 
<div className={formWrapperClass}>
 <form
```

tarunvijwani replied:
This is super helpful. Thank you! I have implemented the suggestion. 👍

ralucaStan reviewed (changes requested):
Hey Tarun, thank you for making the implementation more accessible. I added some extra improvements.

ralucaStan commented on assets/js/base/components/cart-checkout/totals/coupon/index.tsx:
```diff
@@ -63,30 +68,33 @@ export const TotalsCoupon = ( {
 
 	return (
-		<Panel
-			className="wc-block-components-totals-coupon"
```
removing this class is a regression, there might be sites that use it. We still need to keep the class in the DOM. The `<>` below should have this class.

tarunvijwani replied:
Nice catch! I have added the class back to the element.

Commit:
Use classnames to add and remove classes

Improved the code to use classnames to add and remove classes. Also, moved the logic to handle click and submit to separate functions.

ralucaStan reviewed (approved):
Nice work Tarun

Commit:
Merge branch 'trunk' into add/7890-coupon-code-link

tarunvijwani commented:
The `Coupon code` text was changed to `Add a coupon`, resulting in the [failing of an e2e test](https://github.com/woocommerce/woocommerce-blocks/actions/runs/3958593916/jobs/6780350665#step:12:119) as it couldn't find the translated text for `Coupon code`. 


The translation for  `Add a coupon` is not [available on translate.wordpress.org](https://translate.wordpress.org/projects/wp-plugins/woo-gutenberg-products-block/stable/nl/default/?filters%5Bterm%5D=Add+a+coupon&filters%5Bterm_scope%5D=scope_any&filters%5Bstatus%5D=current_or_waiting_or_fuzzy_or_untranslated&filters%5Buser_login%5D=&filter=Apply+Filters&sort%5Bby%5D=priority&sort%5Bhow%5D=desc), so we temporarily changed the text to `Apply a coupon code` in woocommerce/woocommerce-blocks#8254 PR until the next WC Core release because the translation for the text `Apply a coupon code`  is already [available](https://translate.wordpress.org/projects/wp-plugins/woo-gutenberg-products-block/stable/nl/default/?filters%5Bterm%5D=coupon&filters%5Bterm_scope%5D=scope_any&filters%5Bstatus%5D=current_or_waiting_or_fuzzy_or_untranslated&filters%5Buser_login%5D=&filter=Apply+Filters&sort%5Bby%5D=priority&sort%5Bhow%5D=desc).

I have opened the https://github.com/woocommerce/woocommerce-blocks/issues/8255 issue to change the text to `Add a coupon`. I have added a reminder to work on it when we are closer to the WC Core release.

Refer to the slack conversation for more details: p1674155165351969-slack-C02UBB1EPEF
